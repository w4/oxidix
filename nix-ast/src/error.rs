use std::{
    fmt::Display,
    num::{ParseFloatError, ParseIntError},
};

use nix_lexer::{Span, Token, TokenDiscriminants};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum Error {
    #[error("Invalid token")]
    InvalidToken,
    #[error("Unexpected end of file")]
    UnexpectedEndOfFile,
    #[error("Unexpected token {0:?}")]
    UnexpectedTopLevelToken(TokenDiscriminants),
    #[error("Unexpected token {0:?} expected one of {1:?}")]
    UnexpectedToken(TokenDiscriminants, Vec<TokenDiscriminants>),
    #[error("Invalid integer: {0}")]
    ParseInt(ParseIntError),
    #[error("Invalid float: {0}")]
    ParseFloat(ParseFloatError),
    #[error("Duplicate formal function argument: {0}")]
    DuplicateFunctionArgument(String),
    #[error("Attribute `{0}` already defined")]
    AttributeAlreadyDefined(String),
}

impl Error {
    pub fn with_span(self, span: Span) -> SpannedError {
        SpannedError { span, error: self }
    }
}

pub trait WithSpan<T> {
    fn with_span(self, span: Span) -> Result<T, SpannedError>;
}

impl<T> WithSpan<T> for Result<T, Error> {
    fn with_span(self, span: Span) -> Result<T, SpannedError> {
        self.map_err(|e| e.with_span(span))
    }
}

#[derive(Debug)]
pub struct SpannedError {
    pub span: Span,
    pub error: Error,
}

impl Display for SpannedError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.error.fmt(f)?;
        write!(f, " (source: {:?})", self.span)?;
        Ok(())
    }
}

impl std::error::Error for SpannedError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        Some(&self.error)
    }
}

pub trait HandleStreamError<T> {
    fn span_error(self) -> Result<(T, Span), SpannedError>;
}

impl<'a, 'b: 'a> HandleStreamError<&'a Token<'a>> for &'b (Result<Token<'a>, ()>, Span) {
    fn span_error(self) -> Result<(&'a Token<'a>, Span), SpannedError> {
        match self.0.as_ref() {
            Ok(v) => Ok((v, self.1.clone())),
            Err(()) => Err(SpannedError {
                span: self.1.clone(),
                error: Error::InvalidToken,
            }),
        }
    }
}

impl<'a, 'b: 'a> HandleStreamError<&'a Token<'a>> for Option<&'b (Result<Token<'a>, ()>, Span)> {
    fn span_error(self) -> Result<(&'a Token<'a>, Span), SpannedError> {
        match self {
            Some((Ok(v), span)) => Ok((v, span.clone())),
            Some((Err(()), span)) => Err(SpannedError {
                span: span.clone(),
                error: Error::InvalidToken,
            }),
            None => Err(SpannedError {
                span: 0..0,
                error: Error::UnexpectedEndOfFile,
            }),
        }
    }
}

impl<'a> HandleStreamError<Token<'a>> for Option<(Result<Token<'a>, ()>, Span)> {
    fn span_error(self) -> Result<(Token<'a>, Span), SpannedError> {
        match self {
            Some((Ok(v), span)) => Ok((v, span)),
            Some((Err(()), span)) => Err(SpannedError {
                span,
                error: Error::InvalidToken,
            }),
            None => Err(SpannedError {
                span: 0..0,
                error: Error::UnexpectedEndOfFile,
            }),
        }
    }
}
