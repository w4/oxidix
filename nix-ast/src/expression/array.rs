use std::iter::Peekable;

use nix_lexer::{SpannedIter, Token, TokenDiscriminants};

use crate::{Expression, HandleStreamError, SpannedError, parse_primary};

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct ArrayExpression<'a> {
    pub values: Vec<Expression<'a>>,
}

impl<'a> ArrayExpression<'a> {
    pub fn parse(
        stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    ) -> Result<ArrayExpression<'a>, SpannedError> {
        let mut values = Vec::new();

        loop {
            let (token, _) = stream.peek().span_error()?;

            if TokenDiscriminants::from(token) == TokenDiscriminants::SquareBracketClose {
                break;
            }

            values.push(parse_primary(stream)?);
        }

        Ok(ArrayExpression { values })
    }
}
