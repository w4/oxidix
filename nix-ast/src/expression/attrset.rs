use std::{collections::BTreeMap, iter::Peekable};

use nix_lexer::{Span, SpannedIter, Token, TokenDiscriminants};

use crate::{
    BindingName, Error, Expression, HandleStreamError, SpannedError, expect_next_token_or_error,
    parse_binding_name, parse_expression_inner,
};

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct AttrsetExpression<'a> {
    pub recursive: bool,
    #[cfg_attr(feature = "serde", serde(borrow))]
    pub bindings: BTreeMap<Vec<BindingName<'a>>, Expression<'a>>,
}

impl<'a> AttrsetExpression<'a> {
    pub fn parse(stream: &mut Peekable<SpannedIter<'a, Token<'a>>>) -> Result<Self, SpannedError> {
        let mut bindings = BTreeMap::new();

        loop {
            match stream.peek().span_error()? {
                (Token::BraceClose, _) => break,
                (Token::Inherit, _) => parse_inherits(stream, &mut bindings)?,
                (Token::Ident(_) | Token::String(_) | Token::InterpolationStart, span) => {
                    parse_key_value(stream, &mut bindings, span)?
                }
                (v, span) => {
                    return Err(Error::UnexpectedToken(
                        v.into(),
                        vec![TokenDiscriminants::BraceClose, TokenDiscriminants::Ident],
                    )
                    .with_span(span));
                }
            }
        }

        expect_next_token_or_error(stream, TokenDiscriminants::BraceClose)?;

        Ok(AttrsetExpression {
            recursive: false,
            bindings,
        })
    }
}

pub fn parse_key_value<'a>(
    stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    bindings: &mut BTreeMap<Vec<BindingName<'a>>, Expression<'a>>,
    span: Span,
) -> Result<(), SpannedError> {
    let key = parse_binding_name(stream)?;
    let value = parse_expression_inner(stream, u8::MAX)?;
    expect_next_token_or_error(stream, TokenDiscriminants::Semicolon)?;

    if bindings.insert(key.clone(), value).is_some() {
        return Err(Error::AttributeAlreadyDefined(format!("{key:?}")).with_span(span));
    }

    Ok(())
}

pub fn parse_inherits<'a>(
    stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    bindings: &mut BTreeMap<Vec<BindingName<'a>>, Expression<'a>>,
) -> Result<(), SpannedError> {
    expect_next_token_or_error(stream, TokenDiscriminants::Inherit)?;

    match stream.peek().span_error()? {
        (Token::Ident(_), _) => parse_inherits_no_from(stream, bindings)?,
        (Token::BracketOpen, _) => parse_inherits_with_from(stream, bindings)?,
        (Token::Semicolon, _) => {}
        (v, span) => {
            return Err(Error::UnexpectedToken(
                v.into(),
                vec![TokenDiscriminants::Ident, TokenDiscriminants::BracketOpen],
            )
            .with_span(span));
        }
    };

    Ok(())
}

pub fn parse_inherits_no_from<'a>(
    stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    bindings: &mut BTreeMap<Vec<BindingName<'a>>, Expression<'a>>,
) -> Result<(), SpannedError> {
    loop {
        match stream.next().span_error()? {
            (Token::Ident(v), span) => {
                if bindings
                    .insert(vec![BindingName::Value(v)], Expression::Ident(v))
                    .is_some()
                {
                    return Err(Error::AttributeAlreadyDefined(v.to_string()).with_span(span));
                }
            }
            (Token::Semicolon, _) => break,
            (v, span) => {
                return Err(Error::UnexpectedToken(
                    v.into(),
                    vec![TokenDiscriminants::Ident, TokenDiscriminants::Semicolon],
                )
                .with_span(span));
            }
        }
    }

    Ok(())
}

pub fn parse_inherits_with_from<'a>(
    stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    bindings: &mut BTreeMap<Vec<BindingName<'a>>, Expression<'a>>,
) -> Result<(), SpannedError> {
    expect_next_token_or_error(stream, TokenDiscriminants::BracketOpen)?;

    let binding = parse_expression_inner(stream, u8::MAX)?;

    expect_next_token_or_error(stream, TokenDiscriminants::BracketClose)?;

    loop {
        match stream.next().span_error()? {
            (Token::Ident(v), span) => {
                if bindings
                    .insert(vec![BindingName::Value(v)], binding.clone())
                    .is_some()
                {
                    return Err(Error::AttributeAlreadyDefined(v.to_string()).with_span(span));
                }
            }
            (Token::Semicolon, _) => break,
            (v, span) => {
                return Err(Error::UnexpectedToken(
                    v.into(),
                    vec![TokenDiscriminants::Ident, TokenDiscriminants::Semicolon],
                )
                .with_span(span));
            }
        }
    }

    Ok(())
}
