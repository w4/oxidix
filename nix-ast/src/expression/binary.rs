use std::iter::Peekable;

use nix_lexer::{SpannedIter, Token, TokenDiscriminants};

use crate::{
    Error, Expression, HandleStreamError, SpannedError, expect_next_token_or_error,
    parse_expression_inner,
};

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct BinaryExpression<'a> {
    pub operator: BinaryOperator,
    pub left: Expression<'a>,
    pub right: Expression<'a>,
    pub postfix: Option<Expression<'a>>,
}

impl<'a> BinaryExpression<'a> {
    pub fn parse(
        stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
        min_prec: u8,
    ) -> Result<Option<(BinaryOperator, Expression<'a>, Option<Expression<'a>>)>, SpannedError>
    {
        let (token, _) = match stream.peek() {
            Some(token) => token.span_error()?,
            None => return Ok(None),
        };

        let Some(operator) = BinaryOperator::parse(token) else {
            return Ok(None);
        };

        let prec = operator.precedence();

        if prec > min_prec {
            return Ok(None);
        }

        // don't chomp the token for juxtaposition-based function application
        if operator != BinaryOperator::FunctionApplication {
            stream.next();
        }

        let next_prec = match operator.associativity() {
            Some(BinaryAssociativity::Left) => prec.saturating_sub(1),
            Some(BinaryAssociativity::Right) => prec,
            None => u8::MAX,
        };

        let right = match operator {
            BinaryOperator::AttributeSelection => parse_binary_expression(stream, next_prec)?,
            _ => parse_expression_inner(stream, next_prec)?,
        };

        let postfix = match operator {
            BinaryOperator::AttributeSelection => {
                if let Some((Ok(Token::TextOr), _)) = stream.peek() {
                    stream.next();
                    Some(parse_expression_inner(stream, next_prec)?)
                } else {
                    None
                }
            }
            _ => None,
        };

        Ok(Some((operator, right, postfix)))
    }
}

fn parse_binary_expression<'a>(
    stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    next_prec: u8,
) -> Result<Expression<'a>, SpannedError> {
    match stream.next().span_error()? {
        (Token::Ident(v), _) => Ok(Expression::Ident(v)),
        (Token::InterpolationStart, _) => {
            let expr = parse_expression_inner(stream, next_prec)?;
            expect_next_token_or_error(stream, TokenDiscriminants::BraceClose)?;
            Ok(expr)
        }
        (Token::String(_), _) => todo!(),
        (v, span) => Err(Error::UnexpectedToken(
            v.into(),
            vec![
                TokenDiscriminants::Ident,
                TokenDiscriminants::InterpolationStart,
                TokenDiscriminants::String,
                TokenDiscriminants::Or,
            ],
        )
        .with_span(span)),
    }
}

// https://nix.dev/manual/nix/2.29/language/operators.html
#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord, Copy)]
pub enum BinaryOperator {
    AttributeSelection,
    FunctionApplication,
    HasAttribute,
    ListConcatenation,
    Multiplication,
    Division,
    Subtraction,
    Addition,
    Update,
    Lt,
    Le,
    Gt,
    Ge,
    Eq,
    Ne,
    LogicalAnd,
    LogicalOr,
    LogicalImplication,
    LeftAssocPipeOperator,
    RightAssocPipeOperator,
}

impl BinaryOperator {
    pub fn precedence(self) -> u8 {
        match self {
            Self::AttributeSelection => 1,
            Self::FunctionApplication => 2,
            // 3 = unary arithmetic negation
            Self::HasAttribute => 4,
            Self::ListConcatenation => 5,
            Self::Multiplication | Self::Division => 6,
            Self::Subtraction | Self::Addition => 7,
            // 8 = unary logical NOT
            Self::Update => 9,
            Self::Lt | Self::Le | Self::Gt | Self::Ge | Self::Eq | Self::Ne => 10,
            Self::LogicalAnd => 12,
            Self::LogicalOr => 13,
            Self::LogicalImplication => 14,
            Self::LeftAssocPipeOperator | Self::RightAssocPipeOperator => 15,
        }
    }

    pub fn parse(token: &Token<'_>) -> Option<Self> {
        match token {
            Token::Dot => Some(Self::AttributeSelection),
            Token::Question => Some(Self::HasAttribute),
            Token::PlusPlus => Some(Self::ListConcatenation),
            Token::Asterisk => Some(Self::Multiplication),
            Token::Slash => Some(Self::Division),
            Token::Neg => Some(Self::Subtraction),
            Token::Plus => Some(Self::Addition),
            Token::SlashSlash => Some(Self::Update),
            Token::Lt => Some(Self::Lt),
            Token::Le => Some(Self::Le),
            Token::Gt => Some(Self::Gt),
            Token::Ge => Some(Self::Ge),
            Token::DoubleEquals => Some(Self::Eq),
            Token::BangEquals => Some(Self::Ne),
            Token::And => Some(Self::LogicalAnd),
            Token::Or => Some(Self::LogicalOr),
            Token::Arrow => Some(Self::LogicalImplication),
            Token::PipeRight => Some(Self::LeftAssocPipeOperator),
            Token::PipeLeft => Some(Self::RightAssocPipeOperator),
            Token::If
            | Token::Then
            | Token::Else
            | Token::Let
            | Token::In
            | Token::Semicolon
            | Token::BraceClose
            | Token::BracketClose
            | Token::SquareBracketClose
            | Token::Comma => None,
            _ => Some(Self::FunctionApplication),
        }
    }

    pub fn associativity(self) -> Option<BinaryAssociativity> {
        match self {
            Self::AttributeSelection
            | Self::HasAttribute
            | Self::Lt
            | Self::Le
            | Self::Gt
            | Self::Ge
            | Self::Eq
            | Self::Ne => None,
            Self::FunctionApplication
            | Self::Multiplication
            | Self::Division
            | Self::Subtraction
            | Self::Addition
            | Self::LogicalAnd
            | Self::LogicalOr
            | Self::LeftAssocPipeOperator => Some(BinaryAssociativity::Left),
            Self::ListConcatenation
            | Self::Update
            | Self::LogicalImplication
            | Self::RightAssocPipeOperator => Some(BinaryAssociativity::Right),
        }
    }
}

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord, Copy)]
pub enum BinaryAssociativity {
    Left,
    Right,
}
