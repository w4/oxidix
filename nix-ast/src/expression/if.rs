use std::iter::Peekable;

use nix_lexer::{SpannedIter, Token, TokenDiscriminants};

use crate::{Expression, SpannedError, expect_next_token_or_error, parse_expression_inner};

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct IfExpression<'a> {
    pub condition: Expression<'a>,
    pub then: Expression<'a>,
    pub r#else: Expression<'a>,
}

impl<'a> IfExpression<'a> {
    pub fn parse(
        stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    ) -> Result<IfExpression<'a>, SpannedError> {
        let condition = parse_expression_inner(stream, u8::MAX)?;
        expect_next_token_or_error(stream, TokenDiscriminants::Then)?;
        let then = parse_expression_inner(stream, u8::MAX)?;
        expect_next_token_or_error(stream, TokenDiscriminants::Else)?;
        let r#else = parse_expression_inner(stream, u8::MAX)?;

        Ok(IfExpression {
            condition,
            then,
            r#else,
        })
    }
}
