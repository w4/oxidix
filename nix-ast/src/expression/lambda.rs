use std::{collections::BTreeMap, iter::Peekable};

use nix_lexer::{SpannedIter, Token, TokenDiscriminants};

use crate::{
    Error, Expression, HandleStreamError, SpannedError, expect_next_token_or_error,
    parse_expression_inner,
};

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct LambdaExpression<'a> {
    pub parameters: LambdaParameters<'a>,
    pub body: Expression<'a>,
}

impl<'a> LambdaExpression<'a> {
    pub fn parse(
        orig_stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    ) -> Result<Option<LambdaExpression<'a>>, SpannedError> {
        // clone the stream so we can optimistically parse
        let mut stream = orig_stream.clone();

        let Some(parameters) = LambdaParameters::parse(&mut stream).transpose() else {
            return Ok(None);
        };

        let Some((Ok(Token::Colon), _)) = stream.next() else {
            return Ok(None);
        };

        // if we've confirmed we're in a lambda we can return any syntax errors from parsing parameters now
        let parameters = parameters?;

        let body = parse_expression_inner(&mut stream, u8::MAX)?;

        // since we've passed parsing, fast forward the stream that was passed to us
        *orig_stream = stream;

        Ok(Some(LambdaExpression { parameters, body }))
    }
}

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub enum LambdaParameters<'a> {
    Set(LambdaParametersSet<'a>),
    Named(&'a str),
}

impl<'a> LambdaParameters<'a> {
    fn parse(
        stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    ) -> Result<Option<Self>, SpannedError> {
        match stream.next() {
            Some((Ok(Token::BraceOpen), _)) => {
                Ok(LambdaParametersSet::parse(stream)?.map(LambdaParameters::Set))
            }
            Some((Ok(Token::Ident(ident)), _)) => Ok(Some(LambdaParameters::Named(ident))),
            _ => Ok(None),
        }
    }
}

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct LambdaParametersSet<'a> {
    pub inner: BTreeMap<&'a str, LambdaParameter<'a>>,
    pub extra_ignored: bool,
    pub bound: Option<&'a str>,
}

impl<'a> LambdaParametersSet<'a> {
    fn parse(
        stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    ) -> Result<Option<Self>, SpannedError> {
        let mut inner = BTreeMap::new();
        let mut extra_ignored = false;

        loop {
            let (ident, span) = match stream.next().span_error()? {
                (Token::Ident(ident), span) => (ident, span),
                (Token::BraceClose, _) => break,
                (Token::DotDotDot, _) => {
                    extra_ignored = true;
                    expect_next_token_or_error(stream, TokenDiscriminants::BraceClose)?;
                    break;
                }
                (v, span) => {
                    return Err(Error::UnexpectedToken(
                        v.into(),
                        vec![
                            TokenDiscriminants::Ident,
                            TokenDiscriminants::BraceClose,
                            TokenDiscriminants::DotDotDot,
                        ],
                    )
                    .with_span(span));
                }
            };

            let default = if matches!(stream.peek(), Some((Ok(Token::Question), _))) {
                let _token = stream.next();
                let default = parse_expression_inner(stream, u8::MAX)?;
                Some(default)
            } else {
                None
            };

            if inner.insert(ident, LambdaParameter { default }).is_some() {
                return Err(Error::DuplicateFunctionArgument(ident.to_string()).with_span(span));
            }

            match stream.next().span_error()? {
                (Token::BraceClose, _) => break,
                (Token::Comma, _) => {}
                (v, span) => {
                    return Err(Error::UnexpectedToken(
                        v.into(),
                        vec![
                            TokenDiscriminants::Question,
                            TokenDiscriminants::Comma,
                            TokenDiscriminants::BraceClose,
                        ],
                    )
                    .with_span(span));
                }
            }
        }

        let bound = if matches!(stream.peek(), Some((Ok(Token::At), _))) {
            stream.next();

            match stream.next().span_error()? {
                (Token::Ident(ident), _) => Some(ident),
                (v, span) => {
                    return Err(
                        Error::UnexpectedToken(v.into(), vec![TokenDiscriminants::Ident])
                            .with_span(span),
                    );
                }
            }
        } else {
            None
        };

        Ok(Some(LambdaParametersSet {
            inner,
            extra_ignored,
            bound,
        }))
    }
}

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct LambdaParameter<'a> {
    pub default: Option<Expression<'a>>,
}
