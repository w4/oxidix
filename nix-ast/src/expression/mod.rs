mod array;
mod binary;
mod r#if;
mod r#let;
mod unary;

pub use array::ArrayExpression;
pub use binary::BinaryExpression;
pub use r#if::IfExpression;
pub use r#let::LetExpression;
pub use unary::UnaryExpression;
