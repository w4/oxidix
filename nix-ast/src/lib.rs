mod error;
mod expression;

use std::{iter::Peekable, path::Path};

use ariadne::Source;
pub use error::{Error, SpannedError};
use error::{HandleStreamError, WithSpan};
use expression::{
    ArrayExpression, AttrsetExpression, BinaryExpression, IfExpression, LambdaExpression,
    LetExpression, UnaryExpression,
};
use nix_lexer::{Lexer, SpannedIter, Token, TokenDiscriminants};

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub enum Expression<'a> {
    Bool(bool),
    Int(i64),
    Float(WrappedFloat),
    Add(Box<(Expression<'a>, Expression<'a>)>),
    Let(Box<LetExpression<'a>>),
    Ident(&'a str),
    If(Box<IfExpression<'a>>),
    Binary(Box<BinaryExpression<'a>>),
    Unary(Box<UnaryExpression<'a>>),
    Array(ArrayExpression<'a>),
    Comment,
    Lambda(Box<LambdaExpression<'a>>),
    Attrset(AttrsetExpression<'a>),
}

pub fn parse_expression<'a>(stream: Lexer<'a, Token<'a>>) -> Result<Expression<'a>, SpannedError> {
    parse_expression_inner(&mut stream.spanned().peekable(), u8::MAX)
}

fn parse_expression_inner<'a>(
    stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    min_prec: u8,
) -> Result<Expression<'a>, SpannedError> {
    let mut left = if let Some(expr) = UnaryExpression::parse(stream)? {
        Expression::Unary(Box::new(expr))
    } else if let Some(lambda) = LambdaExpression::parse(stream)? {
        Expression::Lambda(Box::new(lambda))
    } else {
        parse_primary(stream)?
    };

    loop {
        let Some((operator, right, postfix)) = BinaryExpression::parse(stream, min_prec)? else {
            break;
        };

        left = Expression::Binary(Box::new(BinaryExpression {
            operator,
            left,
            right,
            postfix,
        }));
    }

    Ok(left)
}

fn parse_primary<'a>(
    stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
) -> Result<Expression<'a>, SpannedError> {
    let (token, span) = stream.next().span_error()?;

    match token {
        Token::Bool(v) => Ok(Expression::Bool(v)),
        Token::Let => Ok(Expression::Let(Box::new(LetExpression::parse(stream)?))),
        Token::If => Ok(Expression::If(Box::new(IfExpression::parse(stream)?))),
        Token::Int(v) => Ok(Expression::Int(
            v.parse().map_err(Error::ParseInt).with_span(span)?,
        )),
        Token::Float(v) => Ok(Expression::Float(WrappedFloat(
            v.parse().map_err(Error::ParseFloat).with_span(span)?,
        ))),
        Token::Ident(v) => Ok(Expression::Ident(v)),
        Token::BracketOpen => wrapped_interpolated(stream, TokenDiscriminants::BracketClose),
        Token::InterpolationStart => wrapped_interpolated(stream, TokenDiscriminants::BraceClose),
        Token::SquareBracketOpen => Ok(Expression::Array(ArrayExpression::parse(stream)?)),
        Token::BraceOpen => Ok(Expression::Attrset(AttrsetExpression::parse(stream)?)),
        Token::Rec => Ok(Expression::Attrset(AttrsetExpression {
            recursive: true,
            ..AttrsetExpression::parse(stream)?
        })),
        Token::Dollar => todo!("dollar"),
        Token::String(_string_tokens) => todo!("str"),
        Token::MultilineString(_string_tokens) => todo!("multi"),
        Token::Throw => todo!("throw"),
        Token::Path(_) => todo!("path"),
        Token::BlockComment | Token::InlineComment => Ok(Expression::Comment),
        v => Err(Error::UnexpectedTopLevelToken(v.into()).with_span(span)),
    }
}

fn wrapped_interpolated<'a>(
    stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    end: TokenDiscriminants,
) -> Result<Expression<'a>, SpannedError> {
    let node = parse_expression_inner(stream, u8::MAX)?;
    expect_next_token_or_error(stream, end)?;
    Ok(node)
}

#[derive(PartialEq, Debug, Clone)]
pub struct WrappedFloat(f64);

impl Eq for WrappedFloat {}

impl PartialOrd for WrappedFloat {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for WrappedFloat {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.0.total_cmp(&other.0)
    }
}

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub enum BindingName<'a> {
    Lazy(Expression<'a>),
    Value(&'a str),
}

fn parse_binding_name<'a>(
    stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
) -> Result<Vec<BindingName<'a>>, SpannedError> {
    let mut name = Vec::new();

    loop {
        match stream.next().span_error()? {
            (Token::Ident(ident), _) => name.push(BindingName::Value(ident)),
            (Token::InterpolationStart, _) => {
                name.push(BindingName::Lazy(parse_expression_inner(stream, u8::MAX)?));
                expect_next_token_or_error(stream, TokenDiscriminants::BraceClose)?;
            }
            (Token::Equals, _) => return Ok(name),
            (token, span) => {
                return Err(Error::UnexpectedToken(
                    token.into(),
                    vec![
                        TokenDiscriminants::Ident,
                        TokenDiscriminants::InterpolationStart,
                        TokenDiscriminants::Equals,
                    ],
                )
                .with_span(span));
            }
        }
    }
}

fn expect_next_token_or_error<'a>(
    stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    kind: TokenDiscriminants,
) -> Result<(), SpannedError> {
    match stream.next().span_error()? {
        (v, _) if TokenDiscriminants::from(&v) == kind => Ok(()),
        (v, span) => Err(Error::UnexpectedToken(v.into(), vec![kind]).with_span(span)),
    }
}

pub fn format_error(error: SpannedError, path: &Path) -> String {
    use ariadne::{Label, Report, ReportKind};

    let mut out = Vec::new();

    Report::build(ReportKind::Error, error.span.clone())
        .with_message("Syntax error")
        .with_label(Label::new(error.span).with_message(error.error))
        .finish()
        .write_for_stdout(
            Source::from(std::fs::read_to_string(path).unwrap()),
            &mut out,
        )
        .unwrap();

    String::from_utf8(out).unwrap()
}

#[cfg(test)]
mod test {
    use std::fs::read_to_string;

    use insta::{assert_debug_snapshot, assert_snapshot, glob};
    use nix_lexer::Logos as _;

    use crate::{format_error, parse_expression};

    #[test]
    fn fixtures() {
        glob!("../fixtures", "*.nix", |path| {
            let input = read_to_string(path).expect("failed to read fixture");
            let lex = nix_lexer::Token::lexer(&input);
            match parse_expression(lex) {
                Ok(v) => assert_debug_snapshot!("fixture", v, &input),
                Err(e) => {
                    assert_snapshot!("fixture", strip_ansi_codes(&format_error(e, path)), &input)
                }
            }
        });
    }

    fn strip_ansi_codes(input: &str) -> String {
        let mut output = String::new();
        let mut chars = input.chars().peekable();

        while let Some(c) = chars.next() {
            if c == '\x1b' {
                if chars.peek() == Some(&'[') {
                    chars.next(); // Skip '['

                    // Skip all characters until a letter (end of ANSI sequence)
                    while let Some(&next) = chars.peek() {
                        if next.is_ascii_alphabetic() {
                            chars.next(); // consume the letter
                            break;
                        } else {
                            chars.next(); // skip numeric or separator
                        }
                    }
                }
            } else {
                output.push(c);
            }
        }

        output
    }
}
