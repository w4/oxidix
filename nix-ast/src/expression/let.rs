use std::{collections::BTreeMap, iter::Peekable};

use nix_lexer::{SpannedIter, Token, TokenDiscriminants};

use crate::{
    BindingName, Error, Expression, HandleStreamError, SpannedError, expect_next_token_or_error,
    parse_binding_name, parse_expression_inner,
};

use super::attrset::parse_inherits;

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct LetExpression<'a> {
    pub bindings: BTreeMap<Vec<BindingName<'a>>, Expression<'a>>,
    pub inner: Expression<'a>,
}

impl<'a> LetExpression<'a> {
    pub fn parse(
        stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    ) -> Result<LetExpression<'a>, SpannedError> {
        let mut bindings = BTreeMap::new();

        let inner = loop {
            match stream.peek().span_error()? {
                (Token::In, _) => {
                    stream.next();
                    break parse_expression_inner(stream, u8::MAX)?;
                }
                (Token::Ident(_), _) => {
                    let name = parse_binding_name(stream)?;
                    let value = parse_expression_inner(stream, u8::MAX)?;
                    expect_next_token_or_error(stream, TokenDiscriminants::Semicolon)?;
                    bindings.insert(name, value);
                }
                (Token::Inherit, _) => {
                    parse_inherits(stream, &mut bindings)?;
                }
                (token, span) => {
                    return Err(Error::UnexpectedToken(
                        token.into(),
                        vec![
                            TokenDiscriminants::In,
                            TokenDiscriminants::Ident,
                            TokenDiscriminants::Inherit,
                        ],
                    )
                    .with_span(span));
                }
            }
        };

        Ok(LetExpression { bindings, inner })
    }
}
