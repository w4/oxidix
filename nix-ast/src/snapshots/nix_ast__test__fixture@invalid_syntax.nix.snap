---
source: nix-ast/src/lib.rs
expression: "{ in }\n"
input_file: nix-ast/fixtures/invalid_syntax.nix
---
Error: Syntax error
   ╭─[ <unknown>:1:3 ]
   │
 1 │ { in }
   │   ─┬  
   │    ╰── Unexpected token In expected one of [BraceClose, Ident]
───╯
